### PR TITLE
Add ML-DSA test vectors (FIPS 204 standard).

### PR DESCRIPTION
**Pull request by gendx**
This PR adds test vectors for ML-DSA, for the [FIPS 204 standard](https://csrc.nist.gov/pubs/fips/204/final). It's a restart of https://github.com/C2SP/wycheproof/pull/112 without the pre-standard versions (round 3, ipd).

These tests aim to cover the following cases:

- Baseline (signing a "Hello world" message).
- Keys and signatures of the wrong length.
- Signature with bit flips.
- Signature hints with a non-canonical encoding:
  - hints that aren't sorted,
  - hints are not strictly sorted (i.e. the same hint is repeated),
  - too many hints (potentially causing a buffer overflow),
  - non-zero padding after the hints.
- Secret keys with the `s1` or `s2` vectors out of the `[-eta, eta]` range.
- Public key with the `t1` component set to zero (allowing trivial forgeries, but the verification algorithm should still accept signatures for this key).
- Boundary conditions in the signature rejection loop (aiming to detect incorrect comparisons):
  - `z_max` equals `gamma1 - beta - 1` and `gamma1 - beta`,
  - `r0_max` equals `gamma2 - beta - 1` and `gamma2 - beta`,
  - `h_ones` equals `omega` and `omega + 1`,
  - in the case of ML-DSA-44, `|ct0|_max` equals `gamma2 - 1` and `gamma2`.
- A "large" number of SHAKE bytes and of SHAKE blocks generated in the `expand_a`, `expand_s`, `rej_ntt_poly` and `rej_bounded_poly` functions.
- Boundary conditions in arithmetic functions:
  - `power_2_round` function: when the remainder (found in `t0`) is equal to 4096 or -4095,
  - `decompose` (via `high_bits` or `low_bits`): when the condition `r_plus - r_0 = q - 1` happens.

Beyond the baseline API, also covered are:
- The `context` value: empty (default), regular length, or context too long.

**What isn't covered:**
- The randomized variant (only the deterministic variant is covered).
- The pre-hased variant "HashML-DSA" (only the regular variant is covered).

Notable difference(s) from the previous pull request (https://github.com/C2SP/wycheproof/pull/112):
- The algorithm now includes the level (e.g. `"algorithm": "ML-DSA-44"`) to make it easier to consume without looking at file names.
- The 32-byte seed used to generate the private key is included (`privateSeed` field) when meaningful and known. It's notably absent for:
  - cases that check for an invalid encoding of the private key (e.g. wrong size, out-of-range values),
  - test vectors of ML-DSA-44 where no seed is known (by construction) for the boundary condition of `|ct0|_max` (note that it's possible to generate similar vectors with a known seed but using a lot of brute-force).

### CODE REVIEW

**Comment by cpu** [comment]
@gendx Was it intentional to put the test vectors in the schema dir? I think we'd probably want them in `testvectors_v1` (?) or another top-level tests directory.

**Comment by gendx** [comment]
@cpu No that was a typo. Moved them to the suitable directory now :)

File: schemas/mldsa_sign_schema.json [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+          "format": "HexBytes",
+          "description": "Encoded ML-DSA private key"
+        },
+        "tests": {
+          "type": "array",
+          "items": {
+            "$ref": "#/definitions/MlDsaSignTestVector"
+          }
+        }
+      }
```
[the remark this answers is not in the document]
Comment (cpu) [reply schemas/mldsa_sign_schema.json]: Handled in my tacked-on commit.

File: schemas/mldsa_sign_schema.json [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+    "MlDsaSignTestGroup": {
+      "type": "object",
+      "properties": {
+        "type": {
+          "enum": ["MlDsaSign"]
+        },
+        "privateSeed": {
+          "type": "string",
+          "format": "HexBytes",
+          "description": "[optional] 32-byte seed that generated the private key (absent if unknown or if the private key is malformed)"
```
[the remark this answers is not in the document]
Comment (cpu) [reply schemas/mldsa_sign_schema.json]: Handled in my tacked-on commit.

File: schemas/mldsa_sign_schema.json [hunk 3]
```diff
@@ -0,0 +1,103 @@
+{
+  "type": "object",
+  "definitions": {
+    "MlDsaSignTestGroup": {
+      "type": "object",
+      "properties": {
```
[the remark this answers is not in the document]
Comment (cpu) [reply schemas/mldsa_sign_schema.json]: Handled in my tacked-on commit.

**Review by cpu — APPROVED** [review]
@gendx I hope you don't mind, but since it has been some time since my last round of feedback I went ahead and pushed a commit that made my requested changes. PTAL. Happy to tweak further if there's something you disagree with.

I've added the source metadata for each testgroup as:
```json
      "source": {
        "name": "github/gendx",
        "version": "0.1"
      },
```

If there's some other repo/affiliation you'd like here LMK.

**Comment by cpu** [comment]
@FiloSottile Do you want to take a look at these before merging?

**Comment by cpu** [comment]
@bwesterb I saw you ref'd these vectors in https://boringssl-review.googlesource.com/c/boringssl/+/79947 Would you be interested in taking a look at the updated branch?

**Comment by cpu** [comment]
> reviewing these test vectors independently

That's what I was thinking. Taking a look at the branch, commenting that the vectors worked for your needs previously, and hopefully that the changes I made since are to your liking as both a downstream user and independent reviewer.

I would find that helpful if you had the time!

**Comment by bwesterb** [comment]
I've [tested them](https://github.com/cloudflare/circl/pull/552) against our implementation in Go. These look great. There is one point of discussion: the FIPS spec (nor LAMPS draft) demand rejection of private keys with coefficients out of range. Do we need to reject them? The private key can be badly formed in more ways which we can detect: we can check whether it's all zeroes, or do statistical tests. I don't think it makes sense to check for any of them _for security_. A more convincing argument is if one implementation would accept them; others reject them; and there are third implementations that generate them.

**Comment by cpu** [comment]
> I've https://github.com/cloudflare/circl/pull/552 against our implementation in Go. These look great.

Awesome, thanks for taking a look. I'm going to merge this PR now on the back of my review & your independent cross-ref. We can of course iterate further if there's any new feedback.

> There is one point of discussion: the FIPS spec (nor LAMPS draft) demand rejection of private keys with coefficients out of range. Do we need to reject them? 

I don't have a strong opinion here as long as the vectors are sufficiently tagged w/ `flags` to let an implementation decide to handle them differently if they chose.

> BTW, these kind of questions are why I pushed for seed-only private keys, but unfortunately we can't have nice things.

Yeahhhhhh :sweat:

**Comment by cpu** [comment]
Thanks again gendx !
